stablehlo/dialect/Base.cpp

```diff
@@ -261,67 +267,116 @@ FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
   return std::make_pair(size, bound);
 }
 
-// TODO(zhouxin) Refactor to better handle errors and return single type
-LogicalResult inferMostSpecificType(
-    Optional<Location> location, TypeRange inputTypes,
-    SmallVectorImpl<Type>& inferredReturnTypes) {
-  SmallVector<RankedTensorType> rankedTypes;
-  for (auto inputType : inputTypes)
-    if (auto rankedType = inputType.dyn_cast<RankedTensorType>())
-      rankedTypes.push_back(rankedType);
-  if (rankedTypes.empty()) {
-    inferredReturnTypes.push_back(inputTypes[0]);
-    return success();
+// Inference rules for conditional branches (lhs/rhs are commutative):
+//       Dim of lhs     Dim of rhs      Infer
+//  c0:  X              X               X
+//  c1:  X              ?               ?
+//  c2:  X              ?, B            ?, max(X, B)
+//  c3:  ?              ?               ?
+//  c4:  ?              ?, B            ?
+//  c5:  ?, B           ?, C            ?, max(B, C)
+FailureOr<std::pair<int64_t, int64_t>> inferLeastSpecificDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound) {
+  bool isLeftStaticDim = !isDynamicDimSize(leftSize);
+  bool isRightStaticDim = !isDynamicDimSize(rightSize);
+  bool isLeftStaticBound = !isDynamicDimSize(leftBound);
+  bool isRightStaticBound = !isDynamicDimSize(rightBound);
+  int64_t inferredSize = ShapedType::kDynamic;
+  int64_t inferredBound = ShapedType::kDynamic;
+
+  if (isLeftStaticDim || isRightStaticDim) {
+    if (isLeftStaticDim && isRightStaticDim) {
+      if (leftSize != rightSize)
+        return emitOptionalError(location, "Mismatched dimension sizes ",
+                                 leftSize, " and ", rightSize, " in dimension ",
+                                 dim);
+      inferredSize = leftSize;
+    } else if (isLeftStaticBound || isRightStaticBound) {
+      int64_t size = isLeftStaticDim ? leftSize : rightSize;
```

Comment:
nit: we should encode the fact here that if isLeftStaticDim then isLeftStaticBound is false so we could have just one condition. Either with tuple or simple if/else.

isLeftStaticDim ? (leftSize, rightBound) : (rightSize, leftBound)

Reply:
That's neat! Changed to
inferredBound = isLeftStaticDim ? std::max(leftSize, rightBound) : std::max(rightSize, leftBound);